foundations/introduction/motivation_and_mindset.md

```diff
@@ -67,6 +67,20 @@ You will inevitably get stuck at some point in the curriculum, perhaps due to a
 *   Take a break: Allow your diffuse learning state to work on the problem.
 *   Ask for help in our [chat](https://discord.gg/fbFCkYabZB): Come prepared with your research. People will be more willing to help you when they can see you have already put effort into trying to figure out the solution on your own.
 
+### A Note on AI Code Generation
+
+As technology advances, we have seen some incredible tools emerge that can help accelerate coding capacity. One particular area that has exploded in popularity lately is the usage of Large Language Models (LLMS) and generative AIs for code completion (like GitHub Co-Pilot) and code generation (like ChatGPT).
+
+While these tools are amazing, learners should be aware of the impacts that using such tools can have on core competency. David Humphrey, a computer science professor, writes about [ChatGPT](https://blog.humphd.org/cheatgpt/) and its potentially negative impacts on core learning. It is a good read about the pitfalls of using generative AI in an educational context.
+
+For learners that are new to programming, tools like ChatGPT or Github Co-pilot can result in some blindspots and gaps in your knowledge.
+1. By using a generative AI, learners may miss the opportunity to discover how something works and why things are done that way.
+1. Asking good questions is a skill in and of itself, and using generative AI to skip asking people (like our helpful Discord community) means that your development of this skill is delayed.
+1. As David Humphrey mentions, the output of generative AI must be closely scrutinized, and learners who are new to programming generally don't have a good over all understanding to be able to determine if the output is good or bad.
+1. Learning how to provide good prompts to an AI tool is a skill in and of itself, but it is a skill that is supplementary to developing foundational programming skills. The Odin curriculum strives to teach that foundational skill.
+
```

Comment:
A few points that I'd love to add to this. Adding text for these thoughts below but please feel free to edit in whatever way you think works. :)

- AI tools are designed to answer questions and are not designed to help you develop research and problem solving skills. If you ask an AI for information, it will give it to you. If you ask a human, they may invite you to share your understanding of the problem and offer guidance on how to discover a solution. 
- Without practice in research and critical thinking, interviews could be a struggle since it is very likely that applicants would not be allowed to use an AI.

Reply:
And I'm kinda' feeling like we can skip on inviting the discussion around prompting AIs. I feel like this will just result in people saying "Well I'll go learn to do that well." 😅

Reply:
Just my two cents.

I think we will always encounter the "but this is the future, old man!" type rebuttal about how "learning how to use AI will be important", etc, etc. I think mentioning something about learning how to prompt AI for useful answers can get ahead of that. It's like with the current WSL policy, we don't invite discussion but it happens anyway.

Open to removing it though, as I don't feel strongly either way. Would love to hear other input from folks on this one.

Reply:
I added the two points that you brought up, but I didn't make any changes to the "AI prompting" line just yet.

Reply:
> * AI tools are designed to answer questions and are not designed to help you develop research and problem solving skills. If you ask an AI for information, it will give it to you. If you ask a human, they may invite you to share your understanding of the problem and offer guidance on how to discover a solution.

Your comment suggests that we are limited to the user inputting questions and AI outputting answers. That's certainly a popular use case, but its far from the only one. Just because a tool can be used to someone's detriment doesn't mean that it always has to be used that way. It's easy enough to go on google/yt for a tutorial on whatever project you're working on, but it would be ridiculous to ban google/yt because of that. You just learn to develop the instinct know which search results will challenge you and help you learn and which won't. I argue that one can develop a similar instinct if they are given high quality instruction on AI usage and have the desire to not cheat their own learning

Examples of alternate use cases that DONT involved simple Q&A:
- Generate alternate solutions to the solution I've provided and give pros and cons of each approach
- Ask me detailed questions about the code I've provided to test my understanding of it
- Simulate an interview for xyz position. Only ask questions and do not give answers
- Analyze my interview answer and tell me how it might be improved
- Give me possible search queries I can use to learn more about xyz topic
- Give me 5 harder versions of the problem that I input without providing the answer

> * Without practice in research and critical thinking, interviews could be a struggle since it is very likely that applicants would not be allowed to use an AI.

I can see how it can definitely hurt research and critical thinking skills if you use it solely in a question answer format. But given the examples above, I think it can actually be a massive aid to interview prep as long as you're not just having it do the whole thing for you

Just as Odin gives the do's and don'ts for asking good questions, I think similar resources should be provided for the do's and don'ts of prompting in AI assisted learning

Reply:
> > * AI tools are designed to answer questions and are not designed to help you develop research and problem solving skills. If you ask an AI for information, it will give it to you. If you ask a human, they may invite you to share your understanding of the problem and offer guidance on how to discover a solution.
> 
> Your comment suggests that we are limited to the user inputting questions and AI outputting answers. That's certainly a popular use case, but its far from the only one. Just because a tool can be used to someone's detriment doesn't mean that it always has to be used that way. It's easy enough to go on google/yt for a tutorial on whatever project you're working on, but it would be ridiculous to ban google/yt because of that. You just learn to develop the instinct know which search results will challenge you and help you learn and which won't. I argue that one can develop a similar instinct if they are given high quality instruction on AI usage and have the desire to not cheat their own learning
> 
> Examples of alternate use cases that DONT involve simple Q&A:
> 
>     * Generate alternate solutions to the solution I've provided and give pros and cons of each approach
> 
>     * Ask me detailed questions about the code I've provided to test my understanding of it
> 
>     * Simulate an interview for xyz position. Only ask questions and do not give answers
> 
>     * Analyze my interview answer and tell me how it might be improved
> 
>     * Give me possible search queries I can use to learn more about xyz topic
> 
>     * Give me 5 harder versions of the problem that I input without providing the answer
> 
> 
> > * Without practice in research and critical thinking, interviews could be a struggle since it is very likely that applicants would not be allowed to use an AI.
> 
> I can see how it can definitely hurt research and critical thinking skills if you use it solely in a question answer format. But given the examples above, I think it can actually be a massive aid to interview prep as long as you're not just having it do the whole thing for you
> 
> Just as Odin gives the do's and don'ts for asking good questions, I think similar resources should be provided for the do's and don'ts of prompting in AI assisted learning

I think you have to keep in mind that this is the third lesson of the curriculum.

While I don't think your comment is wrong, I do believe that nuance on AI usage does not belong so early and perhaps another lesson, later (perhaps in the getting hired section) might be worthwhile.

I personally think that, given where this addition is, the curriculum should be firmly in the "Do not use this tool at this point" camp.

I see it as analogous to introducing how to use React Hooks in the foundational Javascript section. We're getting there, but it's not the appropriate place to introduce it.

Reply:
@Kuanchiliao1 
> if they are given high quality instruction on AI usage

I don't foresee a situation where we would ever develop content around this. And I also don't see a situation where we would link to resources that invite people to learn this. Thanks for your input.

Reply:
> While I don't think your comment is wrong, I do believe that nuance on AI usage does not belong so early and perhaps another lesson later (perhaps in the getting hired section) might be a good place to add in some nuance to how one might use AI in a way that isn't detrimental to learning.
> 
> I personally think that, given where this addition is, the curriculum should be firmly in the "Do not use this tool at this point" camp.
> 
> I see it as analogous to introducing how to use React Hooks in the foundational Javascript section. We're getting there, but it's not the appropriate place to introduce it.

The difference is that React Hooks has an extremely narrow use case compared to effective AI usage. This seems more akin to learning how to ask good questions, which can be used to learn anything and is covered in the assignment right after this. The exception being that you're asking AI instead of humans of course and you also have to learn the skill of filtering AI output

It's pretty clear that we are heading in a direction where major search engines like Bing and Google are going to have AI assistants anyway. I'm sure you'll be able to disable it, but it will be up in everyone's face already and that's just how search engines are probably going to be like going forward. To me that would be roughly the equivalent of saying to a new student "Google right now is too good and you won't be able to handle any bad answers, so we're going to downgrade the entire thing with filters and advanced search settings you should use in every search"

Reply:
> @Kuanchiliao1
> 
> > if they are given high quality instruction on AI usage
> 
> I don't foresee a situation where we would ever develop content around this. And I also don't see a situation where we would link to resources that invite people to learn this. Thanks for your input.

Real questions - Have you ever seen anyone use AI with the intent of learning? Have you ever tried to do so yourself? Do you believe that there is absolutely no way for AI to enhance the learning effectiveness of a new student learning programming?

It's clear that you think that even if there could be some benefit, it's not worth it mentioning. But I do wonder whether you have ever seen any of the possible benefits first hand

Reply:
@Kuanchiliao1 I think the detriment outweighs the benefit. I think a more pressing issue is the idea that learners could grow used to using this as a tool for research and struggle in interviews when they suddenly don’t have it, and don’t have the habits and skills that come with not using it

I appreciate your remarks. Thanks.

Reply:
> > While I don't think your comment is wrong, I do believe that nuance on AI usage does not belong so early and perhaps another lesson later (perhaps in the getting hired section) might be a good place to add in some nuance to how one might use AI in a way that isn't detrimental to learning.
> > I personally think that, given where this addition is, the curriculum should be firmly in the "Do not use this tool at this point" camp.
> > I see it as analogous to introducing how to use React Hooks in the foundational Javascript section. We're getting there, but it's not the appropriate place to introduce it.
> 
> The difference is that React Hooks has an extremely narrow use case compared to effective AI usage. This seems more akin to learning how to ask good questions, which can be used to learn anything and is covered in the assignment right after this. The exception being that you're asking AI instead of humans of course and you also have to learn the skill of filtering AI output
> 
> It's pretty clear that we are heading in a direction where major search engines like Bing and Google are going to have AI assistants anyway. I'm sure you'll be able to disable it, but it will be up in everyone's face already and that's just how search engines are probably going to be like going forward. To me that would be roughly the equivalent of saying to a new student "Google right now is too good and you won't be able to handle any bad answers, so we're going to downgrade the entire thing with filters and advanced search settings you should use in every search"

Overall, I don't disagree with you. In fact, I do believe AI tools provide a lot of value that companies will be sure to capture for their own bottom line. ChatGPT and Github Copilot being "right-ish" a lot of the time is precisely where this value lies, and why I think your point about "learning to prompt an AI" is valid.

All that being said, for me, we have to keep in mind the context of where this content is going. It's the third lesson of an entire curriculum dedicated to building hard and soft skills in the domain of programming. These skills include things like syntax and grammar for HTML, CSS, Javascript, React, Ruby and Rails. It also includes things like soft amplifier skills: research, asking good questions, time management, the ability to stick with a task, and problem solving. The usage of AI tools can shortcircuit the feedback loop that helps to develop these skills.

Let's talk about your Google example. We have a new student that just uses Google to get an answer. That is to say, they are doing research. This is good. But then they discover an answer on Google that puzzles them. So, what are the options?

1. They could ask a real person why this answer is the way it is. They could ask a good question and the human on the other end may prompt them to research another topic, or consider x, y or z about the problem and how they inform the chosen solution.
1. They could ask an AI. They might be able to form an excellent prompt to get a similiar experience as the "ask a human" scenario, but the question here is how likely that is to be the case. They may ask "show me another way to accomplish this" and the AI will dutifully do so. They may even ask "why this way compared to the other way?" and the AI will spit something out. The trouble is, how is the learner here able to discriminate that the AI is providing good information?[^1][^2][^3][^4]

I think the problem is two fold. Firstly, ChatGPT is based on an LLM. LLMs are designed to provide "reasonable completion" of a given text. At the core, this is what ChatGPT and Copilot is doing. Reasonable completion gives us reasonably good results, but it's that last 5% that is the trouble, and likely to cause issues for learners.

Secondly, most learners are probably not equipped with the tools to discern between good and bad for that 5%.

This brings me to the nuance of considering where this lesson is in the greater Odin curriculum. This is the third lesson where I think this nuance of AI and how it can be an excellent tool. Once you have the foundational skills to read the output closely and consider the context of that snippet, whether or not a snippet will provide value to you and whether or not you should change the snippet; that is when AI has general value to programmers. If Odin were to include a discussion on this, it would definitely not be in this particular lesson.

I don't think anyone here is angrily shouting at the clouds saying AI bad period. I can only speak for myself, but my point is much more nuanced than that: for learners early in their journey, the value that AI brings to the table is directly contradictory to the value that the curriculum is trying to provide.

There is no shortcut to learning how to program. If you decide to use AI, you still need to know enough about how to program to figure out if the output is any good. ThePrimeagen recently released a video about Copilot and his usage of it after 6 months.[^5] His points are very prescient for the Odin curriculum as well. In summary (you are welcome to watch the video), he presents a few rules for using Copilot:

1) "Don't do comment driven development" - the essence of this point is if you never write the code, you never learn anything.
2) "Never let copilot write a function" - this is very telling. A senior level engineer with experience basically always deletes what Copilot comes up with.
3) "Define what you want, and let Copilot do smart copy and paste." - I think this is the crux of the argument about "but AI has so much value, we should teach learners about it ASAP!" This feature is, indeed, very powerful, but it comes with the two caveats above.
4) "Make sure you have your types explicit" - this is a high level nuance, and I don't think it's worth discussion this early in the curriculum. More or less, defining types gives AI more clues about what you want to do. This would be akin to "craft better prompts" - but the core trouble here is that folks don't know what they don't know, and it can result in sub-optimal inputs.
5) His final point stands on its own and is worth watching.[^6] If he recommends that juniors don't use the tool, then folks that are "pre-junior" should definitely avoid it. Like Primeagen, I believe that using generative AI is not in the interest of learners, but has a ton of value generating boilerplate.

[^1]: https://www.businessinsider.com/chatgpt-may-make-up-facts-openai-cto-mira-murati-says-2023-2
[^2]: https://www.wired.com/story/openai-chatgpts-most-charming-trick-hides-its-biggest-flaw/
[^3]: https://www.pbs.org/newshour/science/analysis-chatgpt-is-great-at-what-its-designed-to-do-youre-just-using-it-wrong
[^4]: https://news.ycombinator.com/item?id=33841672
[^5]: https://youtu.be/RDd71IUIgpg
[^6]: https://youtu.be/RDd71IUIgpg?t=487

Reply:
> I can only speak for myself, but my point is much more nuanced than that: for learners early in their journey, the value that AI brings to the table is directly contradictory to the value that the curriculum is trying to provide, and has a negative impact on a learner's potential to learn the important things.

agree 100%

Reply:
> I think the problem is two fold. Firstly, ChatGPT is based on an LLM. LLMs are designed to provide "reasonable completion" of a given text. At the core, this is what ChatGPT and Copilot is doing. Reasonable completion gives us reasonably good results, but it's that last 5% that is the trouble, and likely to cause issues for learners.
>

We don't know enough about our own minds to truly say that we're not basically just doing text completions ourselves. I would add that the goal is to provide reasonable completion *comparable to human output and satisfies all users*. An incorrect or misleading answer does not meet the satisfaction criterion. I guess I'm just trying to say that I don't think the mechanism is fundamentally bad and that it will likely only get far more accurate from here on out. It's surely not a perfect system as is, but I suppose we'll cross that bridge when we get there(to high accuracy outputs). 

> Secondly, most learners are probably not equipped with the tools to discern between good and bad for that 5%.
> 
> This brings me to the nuance of considering where this lesson is in the greater Odin curriculum. This is the third lesson where I think this nuance of AI and how it can be an excellent tool does **not** belong. Of course, once you have the foundational skills to read the output closely and consider the context of that snippet, whether or not a snippet will provide value to you and whether or not you should change the snippet; that is when AI has general value to programmers. If Odin were to include a discussion on this, it would definitely not be in this particular lesson.

> 
> I can only speak for myself, but my point is much more nuanced than that: for learners early in their journey, the value that AI brings to the table is directly contradictory to the value that the curriculum is trying to provide, and has a negative impact on a learner's potential to learn the important things.

It seems like we're only really discussing the detriments, and not wanting to think about or weigh that with the benefits. It's very telling to me that Carlos could not or would not provide any answer to the basic questions I asked here: "Have you ever seen anyone use AI with the intent of learning? Have you ever tried to do so yourself? Do you believe that there is absolutely no way for AI to enhance the learning effectiveness of a new student learning programming?". I also want to add this too: "If you think it _is_ possible, what would effective use actually look like for a learner?"

Without knowing about the benefits for learners and weighing that with the detriments, how can we make a proper decision? The question I'd like to answer the most is what kind of positive impact can this technology have for a beginner and how hard is it to achieve that? That seems to be two of our main points of contention. I don't think it "proper use" is as hard to achieve and I think there's more upside to "proper use" than you do

> Let's talk about your Google example. We have a new student that just uses Google to get an answer. That is to say, they are doing research. This is good. But then they discover an answer on Google that puzzles them. So, what are the options?
> 
> 1. They could ask a real person why this answer is the way it is. They could ask a good question and the human on the other end may prompt them to research another topic, or consider x, y or z about the problem and how they inform the chosen solution.
> 2. They could ask an AI. They might be able to form an excellent prompt to get a similar experience as the "ask a human" scenario, but the question here is how likely that is to be the case. They may ask "show me another way to accomplish this" and the AI will dutifully do so. They may even ask "why this way compared to the other way?" and the AI will spit something out. The trouble is, how is the learner here able to discriminate that the AI is providing good information?[1](#user-content-fn-1-afd83b8af7cbabc1142365c104b171cc)[2](#user-content-fn-2-afd83b8af7cbabc1142365c104b171cc)[3](#user-content-fn-3-afd83b8af7cbabc1142365c104b171cc)[4](#user-content-fn-4-afd83b8af7cbabc1142365c104b171cc)

The 3rd and probably more popular option is to perform more Google searches until you find out. Building a mental model around how to spot false information is not a skill unique to AI. For example, mine is roughly that if it's in documentation or in curriculum I can treat it as roughly the source truth, and if there's a consensus on a site like stackoverflow for example then its very likely to be correct. I would treat most other things with a lot more skepticism and I think that's around same level people can and should apply to AI answers without too much difficulty. Interpreting AI answers seems very similar to figuring out what to trust from a coding blog post(outside curriculum), which people do all the time. 

> There is no shortcut to learning how to program. If you decide to use AI, you still need to know enough about how to program to figure out if the output is any good. The blog article linked in the text expresses this point. Heck, ThePrimeagen recently released a video about Copilot and his usage of it after 6 months[5](#user-content-fn-5-afd83b8af7cbabc1142365c104b171cc) and his points are applicable as well.
> 
> In summary, he presents a few rules for using Copilot:
> 
> 1. "Don't do comment driven development" - the essence of this point is if you never write the code, you never learn anything.
> 2. "Never let copilot write a function" - this is very telling. A senior level engineer with experience basically always deletes what Copilot comes up with.
> 3. "Define what you want, and let Copilot do smart copy and paste." - I think this is the crux of the argument about "but AI has so much value, we should teach learners about it ASAP!" This feature is, indeed, very powerful, but it comes with the two caveats above.
> 4. "Make sure you have your types explicit" - this is a high level nuance, and I don't think it's worth discussion this early in the curriculum. More or less, defining types gives AI more clues about what you want to do. This would be akin to "craft better prompts" - but the core trouble here is that folks don't know what they don't know, and it can result in sub-optimal inputs.
> 5. "Should _you_ use it?" - His final point stands on its own and is worth watching.[6](#user-content-fn-6-afd83b8af7cbabc1142365c104b171cc) If he recommends that juniors don't use the tool, then folks that are "pre-junior" should definitely avoid it. Like Primeagen, I believe that using generative AI is not in the interest of learners, but has a ton of value generating and modifying boilerplate. In other words, Copilot rocks for Java developers 🤣.

Yes, I agree that Copilot as it is should almost certainly never be used by beginners for all of the reasons he stated. I think the higher level reason is that with Copilot, you cannot easily decide to disable it's autosuggest features as you are coding and it ends up "thinking" for you, which is really terrible for the learning experience. I myself turned off Copilot for a lot of my coding journey and still carefully restrict my use of it now. Something like that is not comparable to ChatGPT because with ChatGPT, you are able to easily control what questions you ask it without it giving you an unwanted answers to questions you didn't even want to ask. You are able to answer the question yourself then ask what you want later. Further, Copilot almost never explains why it's doing the things it does, it just does them. They are not comparable tools and the value of Copilot as it currently is in the learning context is indeed quite low in my mind

> The conclusion is that Odin is about teaching foundational programming skills, and the curriculum should recommend (even maybe strongly) against using generative AI tools because they can seriously hamper the learning outcomes of folks learning how to program.

I don't necessarily agree but that's a fair conclusion to settle. At the very least, I think it would be beneficial to demonstrate bad uses cases and explain exactly why they might hinder someone's learning. Demonstrate examples of when AI is wrong and exactly how that might be detrimental in the future for a learner. For something this prevalent, I don't think a simple "we highly recommend against using this" is going to stop most learnings from trying

Reply:
> I don't necessarily agree but that's a fair conclusion to settle. At the very least, I think it would be beneficial to demonstrate bad uses cases and explain exactly why they might hinder someone's learning. Demonstrate examples of when AI is wrong and exactly how that might be detrimental in the future for a learner. For something this prevalent, I don't think a simple "we highly recommend against using this" is going to stop most learnings from trying.

I agree that the curriculum saying so doesn't necessarily stop learners from doing something. Heck, we have learners that use WSL or other Linux distributions even though the curriculum doesn't mention them.

The best that the curriculum is express why AI in the context of learning is not optimal. It is up to folks to trust or consider that point of view and decide for themselves if they should follow the curriculum.

The proposed additions suggest reasons why AI may not be advantageous while learning. They are designed as prompts to get learners to consider why the recommendation being made is against generative AI.

If a learner chooses to use generative AI after considering these warnings, then we can only hope they will use the AI in a learning-positive manner. At least with these notes in the curriculum, then learner is aware of the tool and will may check it out in the future.

The curriculum can never sanction anything, as Odin isn't a certificate based program, and there is no one validating and certifying that learners are actually learning anything. So in that same spirit, the changes here make a strong recommendation against using AI. It's the best we can expect of the curriculum, it's up to learners whether they trust the content and the intention and thought put into it or not.